File: 김순석/프로그래머스/pro_조이스틱.py
def solution(name):
    answer = 0
    N = len(name)
    loc = []
    cnt = 0
    ans = 0

    # A가 아닌 알파벳의 인덱스를 loc에 저장
    for i in range(N):
        if name[i] != 'A':
            loc.append(i)
    # loc 비었으면(A로만 이루어진경우) 0 반환
    if not loc:
        return 0

    elif len(loc) == 1:
        if loc[0] <= N // 2:
            ans = loc[0]
        else:
            ans = N - loc[0]

    # 바꿔야할 알파벳 개수가 2 이상이면
    else:
        for i in range(len(loc) - 1):
            ans = min(loc[-1], loc[i] * 2 + len(name[loc[i + 1]:]))


    # 조작할 문자들의 조작 횟수
    for i in loc:
        # 알파벳 순 N 이전의 것이면 A만큼 빼주고
        if ord(name[i]) <= ord('N'):
            answer += ord(name[i]) - ord('A')
        # 아닌 경우(이전 알파벳으로 이동하는 경우)
        else:
            answer += 91 - ord(name[i])
    return answer + ans

File: AHYEON/10.GREEDY/pro_조이스틱.swift
import Foundation
func solution(_ name:String) -> Int {
    var ans = 0
    let name = name.map({$0})

    // up down 계산
    for i in 0..<name.count {
        if name[i] != "A" && name[i] < "N" {
            ans += Int(name[i].asciiValue! - 65)
        } else if name[i] >= "N" {
            ans += Int(91 - name[i].asciiValue!)
        }
    }

    // left right 계산
    var minMove = name.count-1
    for i in 0..<name.count {
        if name[i] != "A" {
            var next = i + 1
            while next < name.count && name[next] == "A" {
                next += 1
            }
            let move = 2 *  i + name.count - next
            minMove = min(move, minMove)
        }
    }
    return ans + minMove
}

